meson.build

```diff
@@ -0,0 +1,150 @@
+        'buildtype=release',
+        'warning_level=3',
+        'b_ndebug=if-release',
+        'b_lto=true',
+        'strip=true',
+        'optimization=0',
+        ],
+    meson_version : '>= 0.50',
+    version : '4.3_pre1',
+)
+
+project_datadir = get_option('datadir') /  meson.project_name()
+
+pymod = import('python')
+py = pymod.find_installation()
+
+py_version = py.language_version()
+if py_version.version_compare('< 3.5')
+  error('Invalid python version!?')
+endif
+
+py_dep = py.dependency(required : true)
+
+install_data('gnofract4d',
+    install_dir: get_option('bindir'),
+)
+
+subdir('fract4d/c')
+
+install_subdir(
+    'fract4d',
+    install_dir : py.get_install_dir(),
+    exclude_directories : ['tests', 'c', 'fract4d_new'],
+)
+
+install_subdir(
+    'fract4d_compiler',
+    install_dir : py.get_install_dir(),
+    exclude_directories : 'tests',
+)
+
+install_subdir(
+    'fract4dgui',
+    install_dir : py.get_install_dir(),
+    exclude_directories : 'tests',
+)
+
+conf_data = configuration_data()
+conf_data.set('app_datadir', get_option('prefix') / project_datadir)
+conf_data.set('version', meson.project_version())
+configure_file(input : '_conf_data.py.in',
+   output : '_conf_data.py',
+   configuration : conf_data,
+   install : true,
+   install_dir: py.get_install_dir() / 'fract4d',
+)
+
+# Application Data
+
+install_subdir(
+    'manual/public',
+    exclude_directories : 'svg',
+    install_dir : project_datadir / 'help',
+    strip_directory : true,
+)
+
+install_subdir(
+    'formulas',
+    install_dir : project_datadir,
+)
+
+install_subdir(
+    'maps',
+    install_dir : project_datadir,
+)
+
+install_data(
+    'pixmaps/improve_now.png',
+    'pixmaps/explorer_mode.png',
+    'pixmaps/mail-forward.png',
+    'pixmaps/draw-brush.png',
+    'pixmaps/face-sad.png',
+    'pixmaps/autozoom.png',
+    'pixmaps/randomize_colors.png',
+    install_dir : project_datadir / 'pixmaps',
+)
+
+# Desktop Environment Data
+
+install_data(
+    'pixmaps/logo/48x48/gnofract4d.png',
+    install_dir: 'share/pixmaps',
+)
+
+icon_sizes = [16, 32, 48, 64, 128, 256]
+icon_template = 'pixmaps/logo/@0@x@0@/gnofract4d.png'
+icon_dir_template = '@0@x@0@/apps/gnofract4d.png'
+icons = []
+icon_dirs = []
+
+foreach s : icon_sizes
+    icons += icon_template.format(s)
+    icon_dirs += icon_dir_template.format(s)
+endforeach
+
+install_data(
+    icons,
+    rename : icon_dirs,
+    install_dir : 'share/icons/hicolor',
+)
+
+install_data(
+    'gnofract4d.desktop',
+    install_dir : 'share/applications',
+)
+
+install_data(
+    'gnofract4d-mime.xml',
+    install_dir : 'share/mime/packages',
+)
+
+# Test
+
+solink_sh = find_program(meson.source_root() / 'solink.sh')
+custom_target('solink',
+    command : [solink_sh, meson.source_root(),
+        meson.build_root()],
+    output : 'none',
+    build_by_default : true,
+    depends: [fract4d_stdlib, fract4dc],
+)
+
+pytest = find_program('pytest')
+test('testing',
```

Comment:
As I'd like to try and take this forward and help you get to the finishing line, I have done a clone and build, however.. when I try running `ninja test` (or `meson test -v` to see what's going wrong), this step fails on account of:
```
ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: --cov=fract4d --cov=fract4dgui --cov=fract4d_compiler
  inifile: /data/Programming/Gnofract4d/meson/pytest.ini
  rootdir: /data/Programming/Gnofract4d/meson

1/1 testing FAIL           0.32s (exit status 4)
```

I run Arch, so this is clearly a change working its way towards production.. Python is 3.8.3, pytest is 5.4.3. I'm not familiar enough with pytest to debug this

Reply:
Maybe arch has pytest3 or pylint3 or something? 

My concern with the meson change is that we are making it harder for someone to just download the source, build and run. Now they have to install ninja and meson (using pip3, the distro one probably isn't recent enough). Would be good to understand what the readme will look like. Perhaps we can keep a setup.py which simplifies the steps...

Another avenue of course is to come up with an easily-consumable binary package. If we can produce that it's less important how easy build-from-source is. I will try and find the time to see how much hassle it is to build a snap image.

Reply:
> As I'd like to try and take this forward and help you get to the finishing line, I have done a clone and build, however.. when I try running `ninja test` (or `meson test -v` to see what's going wrong), this step fails on account of:
> 
> ```
> ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
> pytest: error: unrecognized arguments: --cov=fract4d --cov=fract4dgui --cov=fract4d_compiler
>   inifile: /data/Programming/Gnofract4d/meson/pytest.ini
>   rootdir: /data/Programming/Gnofract4d/meson
> 
> 1/1 testing FAIL           0.32s (exit status 4)
> ```
> 
> I run Arch, so this is clearly a change working its way towards production.. Python is 3.8.3, pytest is 5.4.3. I'm not familiar enough with pytest to debug this

https://pypi.org/project/pytest-cov/ by the look of it. (pytest 5.4.3 working fine here, although I see now I've broken the version tests).

Reply:
@edyoung I think the same argument can be made about cmake as this is not a tool that comes out the box on most distros and even after Debian/Ubuntu's `build-essential` is installed.. however, I think you may find the argument for Meson's inbuilt packaging support to be compelling:

Meson supports destdir on install out the box which makes writing packaging descriptions considerably easier - for example: https://github.com/DX-MON/crunch/blob/master/package/PKGBUILD or in the case of Debian, the hilariously simple https://github.com/DX-MON/crunch/blob/master/package/debian/rules

So I think you'll find building an installable package this way very easy to do and low maintenance. You also get a dist archive out of the process (`meson dist`) that is ready for upload to a GitHub release.

Reply:
Sounds good. I would just like to see what instructions we need to put in README.md to cover the 'install source dist; build; run locally' path.